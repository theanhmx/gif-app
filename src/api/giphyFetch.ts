import { GiphyFetch } from '@giphy/js-fetch-api';
import { DraftGif } from '../type/gif';

const API_KEY = 'TA0egc9FO0OiKVDPqhOheLQM5ro36A5D';

// use @giphy/js-fetch-api to fetch gifs, instantiate with your api key
export const gf = new GiphyFetch(API_KEY);

export const uploadGif = (draftGif: DraftGif) => {
	const formData = new FormData();
	formData.append('api_key', API_KEY);
	formData.append('file', draftGif.file, draftGif.file.name);
	formData.append('tags', draftGif.tags);
	if (draftGif.source_post_url) {
		formData.append('source_post_url', draftGif.source_post_url);
	}
	return fetch('https://upload.giphy.com/v1/gifs/', {
		method: 'POST',
		body: formData,
	});
};
