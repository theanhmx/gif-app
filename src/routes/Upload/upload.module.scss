.upload {
	display: flex;
	flex-direction: column;
	align-items: center;
}

.title {
	color: #a6a6a6;
	font-family: interface, Helvetica Neue, helvetica, sans-serif;
	font-size: 18px;
	letter-spacing: 0.01em;
	text-transform: capitalize;
	-webkit-font-smoothing: antialiased;
	-moz-osx-font-smoothing: grayscale;
	font-weight: 600;
	margin-top: 16px;
	margin-bottom: 16px;
}

.gifs {
	display: flex;
	flex-direction: column;
	gap: 16px;
}

.hiddenInput {
	position: absolute;
	width: 0;
	height: 0;
	overflow: hidden;
}

.upload,
.fileInput {
	margin-top: 16px;
}
