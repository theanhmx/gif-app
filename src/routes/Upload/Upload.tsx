import React, { ChangeEvent, FC, useEffect, useRef, useState } from 'react';

import { DraftGif } from '../../type/gif';
import { uploadGif } from '../../api/giphyFetch';
import { DraftGif as DraftGifComponent, Button } from '../../components';

import styles from './upload.module.scss';
import { Loader } from '@giphy/react-components';

const IMAGE_MAX_SIZE = 5000000; // 5MB

const Upload: FC = () => {
	const [draftGifs, setDraftGifs] = useState<DraftGif[]>([]);
	const imageInputRef = useRef<HTMLInputElement>(null);
	const [responseCounter, setResponseCounter] = useState(0);
	const [loading, setLoading] = useState(false);

	const onChangeImage = () => {
		imageInputRef?.current?.click();
	};

	const onFileChangeCapture = (e: ChangeEvent<HTMLInputElement>) => {
		const file = e.target.files?.[0];

		if (file?.size && file.size > IMAGE_MAX_SIZE) {
			alert('File is too big!');
			// Reset input
			e.target.value = '';
			return;
		}

		if (file !== undefined) {
			setDraftGifs((prev) => [
				...prev,
				{
					file,
					tags: '',
					source_post_url: '',
				},
			]);
		}
	};

	const submit = () => {
		setResponseCounter(0);
		setLoading(true);
		draftGifs
			.filter((gif) => !gif.id) // only upload gifs that have not been uploaded yet (no id)
			.forEach(async (gif) => {
				const response = await uploadGif(gif);
				const { data } = await response.json();
				if (data.id) {
					setDraftGifs((prev) =>
						prev.map((prevGif) => {
							if (prevGif.file === gif.file) {
								return {
									...prevGif,
									id: data.id,
								};
							}
							return prevGif;
						})
					);
				}
				setResponseCounter((prev) => prev + 1);
			});
	};

	useEffect(() => {
		const newGifs = draftGifs.filter((gif) => !gif.id);
		if (responseCounter === newGifs.length) {
			setLoading(false);
		}
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [responseCounter]);

	return (
		<div className={styles.upload}>
			<span className={styles.title}>Upload your GIFs</span>
			<div className={styles.gifs}>
				{draftGifs.map((gif, i) => (
					<DraftGifComponent
						key={i}
						gif={gif}
						onChange={(newGif) => {
							setDraftGifs((prev) =>
								prev.map((prevGif, index) => {
									if (index === i) {
										return newGif;
									}
									return prevGif;
								})
							);
						}}
						onRemove={() => {
							setDraftGifs((prev) => prev.filter((_, index) => index !== i));
						}}
					/>
				))}
			</div>
			<input
				type="file"
				accept="image/gif,video/mp4,video/mov,video/quicktime,video/webm"
				onChangeCapture={onFileChangeCapture}
				name="Add your GIF"
				ref={imageInputRef}
				className={styles.hiddenInput}
				data-testid="file-input"
			/>
			<Button
				text={draftGifs.length === 0 ? 'Choose your GIF' : 'Add more GIF'}
				onClick={onChangeImage}
				className={styles.fileInput}
			/>
			{loading && <Loader />}
			{draftGifs.length > 0 && !loading && (
				<Button
					className={styles.upload}
					onClick={submit}
					text="Submit"
					disabled={loading}
					data-testid="submit"
				/>
			)}
		</div>
	);
};

export default Upload;
