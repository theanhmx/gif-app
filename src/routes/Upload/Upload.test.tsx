/* eslint-disable testing-library/no-wait-for-side-effects */
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import * as GiphyFetch from '../../api/giphyFetch';

import Upload from './Upload';

describe('Upload', () => {
	window.URL.createObjectURL = jest.fn();

	afterEach(() => {
		(window.URL.createObjectURL as any).mockReset();
	});

	it('should render correctly', () => {
		const { container } = render(<Upload />);

		expect(container).toMatchSnapshot();
	});

	it('should render the selected gif and submit button', async () => {
		render(<Upload />);

		const input = screen.getByTestId('file-input');
		const file = new File(['(⌐□_□)'], 'chucknorris.gif', { type: 'image/gif' });
		await waitFor(() =>
			fireEvent.change(input, {
				target: { files: [file] },
			})
		);

		const gif = screen.getByTestId('gif-img');
		const submitButton = screen.getByTestId('submit');
		expect(gif).toBeInTheDocument();
		expect(submitButton).toBeInTheDocument();
	});

	it('should remove the file when remove is clicked', async () => {
		render(<Upload />);

		const input = screen.getByTestId('file-input');
		const file = new File(['(⌐□_□)'], 'chucknorris.gif', { type: 'image/gif' });
		await waitFor(() =>
			fireEvent.change(input, {
				target: { files: [file] },
			})
		);

		const gif = screen.getByTestId('gif-img');
		const removeButton = screen.getByText('Remove');
		fireEvent.click(removeButton);
		expect(gif).not.toBeInTheDocument();
	});

	it('should upload the gif with correct info', async () => {
		const mockUploadGif = jest.spyOn(GiphyFetch, 'uploadGif');
		render(<Upload />);

		const input = screen.getByTestId('file-input');
		const file = new File(['(⌐□_□)'], 'chucknorris.gif', { type: 'image/gif' });
		await waitFor(() =>
			fireEvent.change(input, {
				target: { files: [file] },
			})
		);

		const tagsInput = screen.getByTestId('tag-input') as HTMLInputElement;
		fireEvent.change(tagsInput, { target: { value: 'new tag' } });

		const sourceInput = screen.getByTestId('source-input') as HTMLInputElement;
		fireEvent.change(sourceInput, { target: { value: 'new source' } });

		const submitButton = screen.getByTestId('submit');
		fireEvent.click(submitButton);

		expect(mockUploadGif).toHaveBeenCalledWith({
			file,
			source_post_url: 'new source',
			tags: 'new tag',
		});
	});
});
