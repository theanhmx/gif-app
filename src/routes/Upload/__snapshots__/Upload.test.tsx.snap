// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload should render correctly 1`] = `
<div>
  <div
    class="upload"
  >
    <span
      class="title"
    >
      Upload your GIFs
    </span>
    <div
      class="gifs"
    />
    <input
      accept="image/gif,video/mp4,video/mov,video/quicktime,video/webm"
      class="hiddenInput"
      data-testid="file-input"
      name="Add your GIF"
      type="file"
    />
    <button
      class="button fileInput"
    >
      Choose your GIF
    </button>
  </div>
</div>
`;
